MapleServer2/Servers/Game/DungeonManager.cs

```diff
@@ -0,0 +1,132 @@
+        //RecyclableMapInstanceIds[mapId].Remove(mapInstanceId);
+        //        if (RecyclableMapInstanceIds[mapId].Count == 0)
+        //        {
+        //            RecyclableMapInstanceIds.Remove(mapId);
+        //        }
+        //        return mapInstanceId;
+
+        public DungeonSession CreateDungeonSession(int dungeonId, DungeonType dungeonType)
+        {
+            int dungeonSessionId = GetUniqueSessionId();
+            int dungeonInstanceId = GetMapInstanceId();
+            DungeonSession dungeonSession = new DungeonSession(dungeonSessionId, dungeonId, dungeonInstanceId, dungeonType);
+            DungeonSessionList.Add(dungeonSessionId, dungeonSession);
+            return dungeonSession;
+        }
+
+        public void RemoveDungeonSession(int dungeonSessionId)
+        {
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return;
+            }
+            int currentDungeonSessionId = DungeonSessionList[dungeonSessionId].SessionId;
+            int currentDungeonInstanceId = DungeonSessionList[dungeonSessionId].DungeonInstanceId;
+            RecyclableSessionIds.Add(currentDungeonSessionId);
+            RecyclableMapInstanceIds.Add(currentDungeonInstanceId);
+            DungeonSessionList.Remove(dungeonSessionId);
+        }
+
+        public int GetUniqueSessionId()
+        {
+            if (RecyclableSessionIds.Count > 0)
+            {
+                int sessionId = RecyclableSessionIds.First();
+                RecyclableSessionIds.Remove(sessionId);
+                return sessionId;
+            }
+            return SessionId++;
+        }
+
+        public DungeonSession GetDungeonSession(int dungeonSessionId)
+        {
+            //set dungeon of session id to x on swich dungeon
+            //if dungeonsession exists but id != dungeonid call dungeonid change
+            //update dungeonid or handle it somewhere else
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return null;
+            }
+            return DungeonSessionList[dungeonSessionId];
+        }
+
+        public DungeonSession GetDungeonSessionByInstance(int instanceId)
+        {
+            //TODO: if no dungeonsession is found, return null
+            return DungeonSessionList.FirstOrDefault(session => session.Value.DungeonInstanceId == instanceId).Value;
+        }
+
+        public bool DungeonUsesMap(DungeonSession dungeonSession, FieldManager fieldManager, Player player)
+        {
+            //fieldmanager is the map the player is coming from/that is to be released
+            if (dungeonSession != null) //is not null after entering dungeon via directory
+            {
+                //check map that is left: fieldmanager.mapId is the to be released map
+                if (dungeonSession.ContainsMap(fieldManager.MapId))  //map is a dungeon map
```

Comment:
`dungeonSession.ContainsMap(fieldManager.MapId)` and `dungeonSession.ContainsMap(player.MapId)` isn't always the same value??

Reply:
fieldManager is the instance that Release is given to decide whether it should be destroyed or not. But, we do not want to destroy the instance if a player moves from dungeon to lobby or lobby to dungeon.
We only want to destroy the instance if the player leaves a dungeon map and enters a map that is not a dungeon map.
player.MapId contains the map the player is going to.

Reply:
to give an example:
when going from tria to a dungeon lobby via dungeon directory:
fieldManager.MapId is tria and player.MapId is the dungeon lobby.
Because releasing happens after player.warp has run, and at that point player.MapId is already assigned the map the player will warp to.